### ISSUE TITLE
PyTorch Lightning Flash Integration

### ISSUE THREAD

**Issue by tchaton** [facebookresearch/pytorchvideo#16]
## 🚀 Feature

NOTE: Please look at the existing list of Issues tagged with the label ['enhancement`](https://github.com/facebookresearch/pytorchvideo/issues?q=label%3Aenhancement). **Only open a new issue if you do not see your feature request there**.

## Motivation

Dear people from PyTorchVideo,

First, congratulations on releasing this framework. It is fabulous !

We plan to integrate this framework within Lightning Flash.

We recently created a new ``BETA`` data processing API called DataPipeline. It makes Dataset obsolete and enable very thin customization and quick data augmentation experimentation.

It is built out of Preprocess and Postprocess with multiple hooks to override. It aims at bridging the skew between training / serving.

Here is the tutorial: https://lightning-flash.readthedocs.io/en/latest/custom_task.html

Here is the doc about it: https://lightning-flash.readthedocs.io/en/latest/general/data.html

Here is the Lightning Flash GitHub: https://github.com/PytorchLightning/lightning-flash

We are really keen to collaborate to make this framework integrated within Lightning Flash.

Small nits for the documentation:

* PyTorch Lightning provides quantization as a Callback: https://pytorch-lightning.readthedocs.io/en/stable/advanced/pruning_quantization.html?highlight=Quantization#quantization

* PyTorch Lightning help with jit export: https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.core.lightning.html?highlight=to_jit#pytorch_lightning.core.lightning.LightningModule.to_torchscript

* LightningModule provides a `load_from_checkpoint` function which will download automatically from url and doesn't require to instantiate the model. It relies on `save_hyperparameters` internally.
Here: https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#save-hyperparameters
This should work.
```
model = EfficientX3d.load_from_checkpoint('https://dl.fbaipublicfiles.com/pytorchvideo/model_zoo/kinetics/efficient_x3d_xs_original_form.pyth')
```

Best,
Thomas Chaton.

## Pitch

NOTE: we only consider adding new features if they are useful for many users.

**Comment by haooooooqi** [comment]
Hi Thomas Chaton,

Thanks for playing with PyTorchVideo (PTV) and we are excited to see you interested in integrating PTV into Lightning Flash!
We are interested in the proposal on integrating PTV into Lightning Flash data fetching/ loading part.
After quick reviewing Lightning Flash's API, seems PTV can be integrated with Lightning Flash from different level, (e.g., EncodedVideoDataset, EncodedVideo, etc). Do you want to share your thoughts on which level do you want to integrate PTV into Lightning Flash?

Thanks,
Haoqi

**Comment by tchaton** [comment]
Dear Haoqi,

Thanks for the quick answer ! That's really exciting !

From the documentation, I can see 2 first tasks to kick-off the integration: `VideoClassifier` and `VideoEmbedder`.

This would require introducing the followings objects: 

Classification:
- VideoClassifier
- VideoClassificationDataModule
- VideoClassificationPreprocess and VideoClassificationPostprocess

Embedder:
- VideoEmbedder
- VideoEmbedderDataModule
- VideoEmbedderPreprocess and VideoEmbedderPostprocess

I assume the ``VideoEmbdder`` to re-use most of ``VideoClassifier`` code.

We can also support ``EncodedVideoDataset`` directly, but it is a perfect opportunity for us to start stressing out the DataPipeline API.

We will have Lightning 1.3 release at the end of next sprint, so we might start this collaboration on the following one.

How does it sound ?

Also, would it be possible for you and your team to join PyTorch Lightning Slack, so we can coordinate there: https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A

Best,
Thomas Chaton.

**Comment by tullie** [comment]
This sounds great @tchaton. I've been working on a VideoClassification LightingModule example which I can submit this afternoon. It might be a good place to start for a Flash module. I'll ping you on the Lighting chat and we can discuss more.

**Comment by tchaton** [comment]
Hey @tullie @haooooooqi,

Here is the draft PR: https://github.com/PyTorchLightning/lightning-flash/pull/216/files. 

I would say 80 % of the `VideoClassifier` is done.

TODOS:
- [] Add support for transforms
- [] Resolve issues with model forwarding
- [] Add tests
- [] Restructure code architecture.

If I can your help to debug, add tests, refine API.
We could finalise this integration early next week and respectively communicate about it on social :)

Best,
Thomas Chaton.

**Comment by kalyanvasudev** [comment]
Closing the issue since the feature has been merged in Flash.
